reconcile/openshift_base.py

```diff
@@ -135,94 +149,109 @@ def init_specs_to_fetch(
+                        resource_names=managed_names,
+                    )
                 )
-                for t in managed_types
-            )
 
             # Initialize desired state specs
             openshift_resources = namespace_info.get("openshiftResources")
             for openshift_resource in openshift_resources or []:
-                d_spec = StateSpec(
-                    "desired",
-                    oc,
-                    cluster,
-                    namespace,
-                    openshift_resource,
-                    namespace_info,
-                    privileged=privileged,
+                state_specs.append(
+                    DesiredStateSpec(
+                        oc=oc,
+                        cluster=cluster,
+                        namespace=namespace,
+                        resource=openshift_resource,
+                        parent=namespace_info,
+                        privileged=privileged,
+                    )
                 )
-                state_specs.append(d_spec)
+
     elif clusters:
         # set namespace to something indicative
         namespace = "cluster"
         for cluster_info in clusters:
             cluster = cluster_info["name"]
-            oc = oc_map.get(cluster)
-            if not oc:
-                if oc.log_level >= logging.ERROR:
+            try:
+                oc = oc_map.get_cluster(cluster)
+            except OCLogMsg as ex:
+                if ex.log_level >= logging.ERROR:
                     ri.register_error()
-                logging.log(level=oc.log_level, msg=oc.message)
+                logging.log(level=ex.log_level, msg=ex.message)
                 continue
 
             # we currently only use override_managed_types,
             # and not allow a `managedResourcesTypes` field in a cluster file
-            for resource_type in override_managed_types or []:
-                ri.initialize_resource_type(cluster, namespace, resource_type)
+            for kind in override_managed_types or []:
+                ri.initialize_resource_type(cluster, namespace, kind)
+
                 # Initialize current state specs
-                c_spec = StateSpec("current", oc, cluster, namespace, resource_type)
-                state_specs.append(c_spec)
+                state_specs.append(
+                    CurrentStateSpec(
+                        oc=oc,
+                        cluster=cluster,
+                        namespace=namespace,
+                        kind=kind,
+                        resource_names=[],
+                    )
+                )
                 # Initialize desired state specs
-                d_spec = StateSpec("desired", oc, cluster, namespace, resource_type)
-                state_specs.append(d_spec)
+                # it seems this StateSpec has no effect and can be disabled. the only code path
+                # leading to this place is from openshift_clusterrolebindings > ob.fetch_current_state >
```

Comment:
This is some serious and brave digging. I really hope we can kill this and make the `cluster` sub-branch trivial.